sw/device/tests/sram_ctrl_execution_test_main.c

```diff
@@ -0,0 +1,117 @@
+     *   +-----------+---------+---------+-------+--------+-----------+
+     *   |  0000000  |  01010  |  01011  |  000  | 01010  |  0110011  |
+     *   +-----------+---------+---------+-------+--------+-----------+
+     *   31        25 24     20 19     15 14   12 11     7 6          0
+     */
+    0x00A58533,
+    /*
+     *   Abstract representation: ret == jalr rd, offset(rs1)
+     *   +----------------+-------+-------+------+-----------+
+     *   |  000000000000  |  rs1  |  000  |  rd  |  1100111  |
+     *   +----------------+-------+-------+------+-----------+
+     *   31             20 19   15 14   12 11   7 6          0
+     *
+     *   Custom representation: jalr zero, 0(ra)
+     *   +----------------+---------+-------+---------+-----------+
+     *   |  000000000000  |  00001  |  000  |  00000  |  1100111  |
+     *   +----------------+---------+-------+---------+-----------+
+     *   31             20 19     15 14   12 11      7 6          0
+     */
+    0x00008067,
+};
+
+/*
+ * Adds the first parameter to the second parameter and returns the result.
+ *
+ * Maps to the `execution_test_instructions` buffer as the function body.
+ */
+typedef uint32_t (*ram_exec_function_t)(volatile uint32_t, volatile uint32_t);
+
+bool test_main(void) {
+  sram_ctrl_t sram = sram_ctrl_main_init();
+
+  bool locked;
+  CHECK_DIF_OK(
+      dif_sram_ctrl_is_locked(&sram.dif, kDifSramCtrlLockExec, &locked));
+  CHECK(!locked, "Execution from SRAM is locked, cannot perform the test");
+
+  dif_toggle_t enabled;
+  CHECK_DIF_OK(dif_sram_ctrl_exec_get_enabled(&sram.dif, &enabled));
+  if (enabled == kDifToggleDisabled) {
+    CHECK_DIF_OK(dif_sram_ctrl_exec_set_enabled(&sram.dif, kDifToggleEnabled));
+  }
+
+  // Map the `execution_test_instructions` to the `ram_exec_function_t`. In
+  // other words - tell the compiler to treat the buffer as the function body.
+  ram_exec_function_t func = (ram_exec_function_t)execution_test_instructions;
+
+  LOG_INFO("Executing from SRAM at = %x", func);
+  LOG_INFO("Instruction[0] = %x (add a0, a1, a0), at %x",
+           execution_test_instructions[0], &execution_test_instructions[0]);
+  LOG_INFO("Instruction[1] = %x (jalr ra, x0) at %x",
+           execution_test_instructions[1], &execution_test_instructions[1]);
+
+  uint32_t expected_result = kExecParam1 + kExecParam2;
+  uint32_t result = func(kExecParam1, kExecParam2);
+  CHECK(result == expected_result,
+        "SRAM Execution expected %d + %d = %d, got %d", kExecParam1,
+        kExecParam2, expected_result, result);
+
+  dif_toggle_t state;
+  CHECK_DIF_OK(dif_sram_ctrl_exec_get_enabled(&sram.dif, &state));
+  CHECK(state == kDifToggleEnabled, "Execution must be enabled");
+  CHECK_DIF_OK(dif_sram_ctrl_exec_set_enabled(&sram.dif, kDifToggleDisabled));
+  CHECK_DIF_OK(dif_sram_ctrl_exec_get_enabled(&sram.dif, &state));
+  CHECK(state == kDifToggleDisabled, "Execution must be disabled");
+
+  result = func(kExecParam1, kExecParam2);
```

Comment:
This comment is related to the previous revision, and is no longer applicable. The current revision does not have a "negative" test (it is expected that the EXEC is enabled, or can be enabled, otherwise test fails).

Reply:
The negative test and refactor I was planning on doing in subsequent PRs, but it would be easier to do if this is merged.